python-spec/src/somacore/coordinates.py

```diff
@@ -0,0 +1,491 @@
+        return tuple(axis.name for axis in self._axes)
+
+
+class CoordinateTransform(metaclass=abc.ABCMeta):
+    """A coordinate transformation from one coordinate space to another.
+
+    Args:
+        input_axes: The names of the axes for the input coordinate space.
+        output_axes: The names of the axes for the output coordinate space.
+
+    CoordinateTransform classes are composable using the ``@`` (__matmul__) operator.
+
+    Lifecycle: experimental
+    """
+
+    def __init__(
+        self,
+        input_axes: Union[str, Sequence[str]],
+        output_axes: Union[str, Sequence[str]],
+    ):
+        self._input_axes = (
+            (input_axes,) if isinstance(input_axes, str) else tuple(input_axes)
+        )
+        self._output_axes = (
+            (output_axes,) if isinstance(output_axes, str) else tuple(output_axes)
+        )
+
+    @abc.abstractmethod
+    def __matmul__(self, other: "CoordinateTransform") -> "CoordinateTransform":
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def __rmatmul__(self, other: "CoordinateTransform") -> "CoordinateTransform":
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def inverse_transform(self) -> "CoordinateTransform":
+        """Returns the inverse coordinate transform.
+
+        Lifecycle: experimental
+        """
+        raise NotImplementedError()
+
+    @property
+    def input_axes(self) -> Tuple[str, ...]:
+        """The names of the axes of the input coordinate space.
+
+        Lifecycle: experimental
+        """
+        return self._input_axes
+
+    @property
+    def output_axes(self) -> Tuple[str, ...]:
+        """The names of the axes of the output coordinate space.
+
+        Lifecycle: experimental
+        """
+        return self._output_axes
+
+
+class AffineTransform(CoordinateTransform):
+    """An affine coordinate trasformation from one coordinate space to another.
+
+    An affine transform is a combination of a linear transformation and a translation.
+
+    Args:
+        input_axes: The names of the axes for the input coordinate space.
+        output_axes: The names of the axes for the output coordinate space.
+        matrix: Matrix (perhaps augmented) that represents the affine transformation.
+            Can be provided as just the linear transform (if no translation), the
+            full augmented matrix, or the augmented matrix without the final row.
+
+    Lifecycle: experimental
+    """
+
+    def __init__(
+        self,
+        input_axes: Union[str, Sequence[str]],
+        output_axes: Union[str, Sequence[str]],
+        matrix: npt.ArrayLike,
+    ):
+        super().__init__(input_axes, output_axes)
+
+        # Check the rank of the input/output axes match.
+        if len(self.input_axes) != len(self.output_axes):
+            raise ValueError(
+                "The input axes and output axes must be the same length for an "
+                "affine transform."
+            )
```

Comment:
It looks like this is true for all types of transform. Should this check be moved to the base `CoordinateTransform`?

Reply:
No, at some point we will want to add in transforms the map between spaces with different number of dimensions. This will be important for 3D support in the long run.

Reply:
Makes sense. It might be worth adding something like

```python
def _check_same_length(self) -> None:
  if len(self.input_axes) != len(self.output_axes):
    raise ValueError(
      f"The input axes and output axes of a {type(self).__name__} must be the same length."
    )
```

so that you can add `self._check_same_length` into only the initializers that need it.

Reply:
bump on the idea above, to reduce the duplication in these `__init__`s (and make the errors nice and consistent)

Reply:
I intentionally left that inline since it's only called in the one place now that I'm using proper inheritance. Do you think it should still create a new function anyway?